src/averageAges.js

```diff
@@ -1,64 +1,74 @@
 'use strict';
 
 /**
- * Implement calculateMenAverageAge function
- *
- * Function returns average age of men in array. If `century` is specified then
- * function calculates average age only for men who died in this century
- *
- * To calculate century:
- * Divide year of person's death by 100: Math.ceil(person.died / 100)
- *
  * @param {object[]} people
  * @param {number} century - optional
  *
  * @return {number}
  */
+
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  const isMale = person => person.sex === 'm';
+
+  const man = people.filter((person) =>
+    isMale(person) && (!century
+    || Math.ceil(person.died / 100) === century)
+  );
+
+  const totalMenAge = man.reduce((sum, person) =>
+    sum + (person.died - person.born), 0
+  );
+
+  return man.length ? totalMenAge / man.length : 0;
 }
 
 /**
- * Implement calculateWomenAverageAge function
- *
- * Function returns average age of women in array. If `withChildren` is
- * specified then function calculates average age only for women with children
- *
- * Hint: To check if a woman has children you should find someone who mention
- * her as mother.
- *
  * @param {object[]} people
  * @param {boolean} withChildren - optional
  *
  * @return {number}
  */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const women = people.filter((person) =>
+    person.sex === 'f' && (!withChildren
+    || people.some((child) => child.mother === person.name))
+  );
+
+  const totalWomenAge = women.reduce((sum, person) =>
+    sum + (person.died - person.born), 0
+  );
+
+  return women.length ? totalWomenAge / women.length : 0;
 }
 
 /**
- * Implement calculateAverageAgeDiff function.
- *
- * The function returns an average age difference between a child and his or her
- * mother in the array. (A mother's age at child birth)
- *
- * If `onlyWithSon` is specified then function calculates age difference only
- * for sons and their mothers.
- *
  * @param {object[]} people
  * @param {boolean} onlyWithSon - optional
  *
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // 1. find a mother of each person (or only for men)
-  // 2. keep people who have mothers in the array
-  // 3. calculate the difference child.born - mother.born
-  // 4. return the average value
+  const isMale = person => person.sex === 'm';
+
+  const relevantPeople = onlyWithSon
+    ? people.filter(person => isMale(person))
+    : people;
+
+  const ageDifferences = relevantPeople.map(person => {
+    const mother = people.find(candidateMother =>
+      candidateMother.name === person.mother
+    );
+
+    return mother
+      ? person.born - mother.born
+      : null;
+  }).filter(ageDiff => ageDiff !== null);
```

Comment:
[CODE STYLE] - All chained methods called on the same level should have same indentation. This makes code much more readable.
You don't need to compare with null, you can just return ageDiff in the filter and it will check if the value exists.